Solution to oop test

wokePanda opened the pull request:
# test_oop

This is a solution to the oop test. 

The program is supposed to take the user's location and a CSV file with all the coffee shops and provide a list of the 3 closest ones.

## Installation

Use [Homebrew](http://brew.sh/) to install Kotlin.

```bash
$ brew update
$ brew install kotlin
```

## Usage

The project can be run from the command line.

### Command line

In order to compile the project navigate to the project folder and run the following command:

```bash
$ kotlinc src -include-runtime -d test_oop.jar
```
This will create a file named: ``` test_oop.jar ```

In order to run the application type in the following command:

```bash
$ java -jar test_oop.jar <user x coordinate> <user y coordinate> <shop data filename>
```

Make sure that the ```.csv``` file is located in the project folder.

wokePanda committed:
Added gitignore

wokePanda committed:
Added project files and main class

- Created the project
- Added the main class and a basic test for it

wokePanda committed:
Parsed input

- Added try block to parse the input
- Added another test for the situation when the user input may not be in the correct format or order

wokePanda committed:
Added .idea folder to gitignore

wokePanda committed:
Created Coordinates class and CoffeeShopFinder class

- Added Coordinates class with tests for it
- Added basic CoffeeShopFinder class

wokePanda committed:
Added DistanceCalculator class

- Created DistanceCalculator class which includes a function to calculate distance between 2 points
- Added tests for DistanceCalculator

wokePanda committed:
Added tests and refactoring

- Added tests to CoffeeShopFinder
- Changed the names of some of the tests of CoffeeShopLocator

wokePanda committed:
Created CoffeeShop Model and Test

wokePanda committed:
Added CSV line Converter class

- Added class to convert CSV line to CoffeeShop object
- Added tests for the Converter class

wokePanda committed:
Added more tests to CoffeeShopFinderTest

- Added extra CSV files of different kinds for testing
- Added extra test functions to CoffeeShopFinderTest
- Added an exception for cases where the CSV file might be empty

wokePanda committed:
Added more tests to Converter and refactoring

- Added another test to the Converter class
- Created a function that converts a list of CSV lines to a list of CoffeShop objects
- Refactored a test in the Converter class

wokePanda committed:
Renamed the main class from CoffeeShopLocator to Main

wokePanda committed:
Added function to sort coffee shop list by distance and refactoring

- Renamed DistanceCalculator to DistanceUtility
- Added another test and function to DistanceUtility that sorts coffee shops from a list according to their distance to a specified location

wokePanda committed:
Converted CoffeeShop list to String

- Added function with test, that converts a list of CoffeeShops to a multi-line string
- Fixed bug where input was processed incorrectly (latitude and longitude were switched)

wokePanda committed:
Reorganized project files

wokePanda committed:
Refactoring

- Converted Converter object to class
- Converted DistanceUtility object to class

wokePanda committed:
Added more details to error

wokePanda committed:
Turned Coordinates and CoffeeShop into data classes

wokePanda committed:
Created test for CoffeeShop function to create final line string

wokePanda committed:
Added .toString() function to CoffeeShop class

- Added function .toString() to CoffeeShop class that overrides the existing function
- Added exception for the case where distance is not calculated when the .toString() function is called
- Created another test for the CoffeeShop class for the case when distance is null

alexandru-calinoiu commented on tests/Model/CoffeeShopTest.kt:
```diff
@@ -0,0 +1,32 @@
+    @Test
+    fun createCoffeeShop() {
+        val coffeeShop = CoffeeShop(Coordinates(42.4,143.234), "Starbucks")
+        assertEquals("Starbucks", coffeeShop.name)
```
This assertion adds no value, you are testing that property setting working in kotlin, this is a language feature that the kotlin developers should have tested.

alexandru-calinoiu commented on tests/Model/CoffeeShopTest.kt:
```diff
@@ -0,0 +1,32 @@
+        val coffeeShop = CoffeeShop(Coordinates(42.4,143.234), "Starbucks")
+        assertEquals("Starbucks", coffeeShop.name)
+        assertEquals(42.4, coffeeShop.coordinates.latitude)
+        assertEquals(143.234, coffeeShop.coordinates.longitude)
```
The above comments applies to this assertion also but there are a couple of other issues with it:
- you are testing the implementation of another class, it should not be the concern of this test
- this assertion can be done like `assertEquals(Coordinates(42.4, 146.234), coffeeShop.coordinates)` can you tell me why?

wokePanda replied:
Because the ```Coordintes``` class is a data class, meaning it has an autogenerated implementation of ```equals```.

alexandru-calinoiu commented on tests/Helpers/ConverterTest.kt:
```diff
@@ -0,0 +1,50 @@
+internal class ConverterTest {
+
+    @Test
+    fun convertLineToCoffeeShop() {
```
The naming of the tests is not that great, this guy has a nice [blog post](http://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html) about it.

alexandru-calinoiu commented on tests/Helpers/ConverterTest.kt:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Test
+    fun convertListToString() {
```
This is a very confusing test, not sure what is used for.

wokePanda replied:
Here I wanted to test the ```listToString``` function of the ```Converter``` class.

Basically I wanted to make sure that if I pass a list of ```CoffeeShop``` objects to this function, along with a number of top results (```closestCount```) I get a string with the format requested for the output of the program.

alexandru-calinoiu commented on src/Model/CoffeeShop.kt:
```diff
@@ -0,0 +1,14 @@
+
+internal data class CoffeeShop(val coordinates: Coordinates, val name: String) {
+    var distanceFromUser: Double? = null
```
Mutability is considered by many not a good idea, can you tell me a couple of reasons why is that?

wokePanda replied:
- It can lead to unintended side-effects;
- It's not thread safe. As the object can be changed, multiple threads might try to update it at the same time, or they might see separate versions of it;
- It is easier to think about immutable objects since you don't have to think about the ways they might change over time (this can help in debugging for example).

alexandru-calinoiu commented on src/Helpers/DistanceUtility.kt:
```diff
@@ -0,0 +1,15 @@
+
+internal class DistanceUtility {
+
+    fun calculateDistance(from: Coordinates, to: Coordinates): Double {
```
This method should be made private, why do you think is a good idea to limit the public api of a class?

wokePanda replied:
In the context of OOP it's a good idea since you can break up tasks into smaller parts. This way if you have a function that is used internally by the class (to serve another function) then you don't want other classes to have access to it (or to use it). An example of this could be a function that changes the value of an object inside the class, this function can be called inside the class by other functions (as an effect of some other event), but you don't want other classes to access it directly.

alexandru-calinoiu reviewed (commented):
This is a good start,

- oop solution is very basic, would have loved to see some stronger separation of concerns, layers like data (cvs reading), business (closes shop calculations) should have been separated by interfaces so they can easily be tested in isolation
- test are a mix of unit and integration test and have some issues

wokePanda committed:
Created CSVParser class and tests for it

- Added tests for CSVParser class
- Created a special class for parsing a csv file (CSVParser)

wokePanda committed:
Added test to CSVParserTest

wokePanda committed:
Refactored CoffeeShop class

- made the "userLocation" variable immutable
- removed pointless tests

wokePanda committed:
Removed DistanceError

wokePanda committed:
Removed iteration in "sortByDistance" function of DistanceUtility class

wokePanda committed:
Renamed the DistanceUtility tests

wokePanda committed:
Added test and condition to CSVParser

- Added test to CSVParser to check if the input file is empty
- Added condition to the "parseCsvFileWithName" function of CSVParser to check if the file is empty

wokePanda committed:
Renamed CoffeeShopFinder tests and removed useless ones

wokePanda committed:
Removed useless test for Coordinates class

wokePanda committed:
Adapted Converter class

- Adapted Converter class functions to new CoffeeShop format
- Renamed tests of Converter class
- Updated Converter class tests

wokePanda committed:
Refactored CofeeShopTest

wokePanda committed:
Refactored MainKtTest and DistanceUtilityTest

- Renamed the tests
- Removed useless tests

wokePanda committed:
Adapted CoffeeShopFinder and refactoring

wokePanda committed:
Removed "location" parameter from "sortByDistance" function of DistanceUtility

wokePanda committed:
Renamed some of the tests

levelboy commented:
Already talked this thoruhg. Closing. Thank you
